src/aligner/smithwatermanlinear.cpp

```diff
@@ -0,0 +1,112 @@
+#include "smithwatermanlinear.h"
+
+#include <functional>
+#include <iostream>
+
+#include "../similaritymatrix/similaritymatrix.h"
+
+typedef boost::multi_array<double, 2> array_type;
+typedef array_type::index index;
+typedef boost::array<index, 2> index_tupel;
+
+SWAligner::SWAligner(const std::string& first_sequence, const std::string& second_sequence) :
+    sequence_x(first_sequence), sequence_y(second_sequence), similarity_matrix(first_sequence, second_sequence) {}
+
+void SWAligner::traceback(index_tupel idx, unsigned int& preliminary_pos) {
+    const array_type& matrix = similarity_matrix.get_matrix();
+
+    // stopping criterion
+    index index_x = idx[0];
+    index index_y = idx[1];
+    if(matrix[index_x-1][index_y-1] == 0) {
+        sequence_x.insert(index_x-1, "(");
+        sequence_y.insert(index_y-1, "(");
+
+        pos = index_y;
+        return;
+    }
+
+    double n1 = matrix[index_x-1][index_y-1];
+    double n2 = matrix[index_x][index_y-1];
+    double n3 = matrix[index_x-1][index_y];
+
+    // move north-west
+    if((n1 >= n2) && (n1 >= n3)) {
+        traceback({{index_x-1, index_y-1}}, preliminary_pos);
+    }
+    // move west
+    else if ((n2 >= n1) && (n2 >= n3)) {
+        sequence_x.insert(index_x, "-");
+        traceback({{index_x, index_y-1}}, preliminary_pos);
+    }
+    // move north
+    else {
+        sequence_y.insert(index_y, "-");
+        traceback({{index_x-1, index_y}}, preliminary_pos);
+    }
+}
+
+double scoring_function(const char& a, const char& b) {
```

Comment:
Maybe it is better to make it a lambda function as a private member field under `SWAligner`. Same for `on_each_iteration`, so that we don't need `std::bind`.